### DISCUSSION TITLE
Async webhooks are fired twice on an Async event (e.g. PRODUCT_UPDATED event)

### DISCUSSION THREAD

**Discussion by satvikt** [saleor/saleor#9946] [General] [answered]
Saleor async webhooks (asynchronous/asynchronous subscription-based) tend to 
1. fire twice per event, 
2. with the same payload. 

This is not an occasional observation, this is what happens **always**, and for **any** type of async event like PRODUCT_UPDATED or PRODUCT_CREATED.

For example, a webhook based upon the PRODUCT_UPDATED event is fired twice upon updating a product, carrying the same payload.
This causes an issue when we are ingesting the payload in an external, non-idempotent system, wherein the first request leads to 200 OK/201 created while the subsequent one leads to 409 Conflict.

The following database entries from the table `core_eventpayload` show 2 payload entries for a single webhook event. Please note that `issuedAt` is same for both,
![Selection_003](https://user-images.githubusercontent.com/21122115/172567607-adda8668-28b6-41fd-bd02-ec6c79a11d28.png)

I'm not sure I am missing anything here.
If not, is there any specific reason why async webhooks are designed this way?

------------
UPDATE (9 Jul 2022)

I incorporated the product_variant async webhoooks as well. Below are the webhooks fired, in the exact sequence they are fired:

1. Upon product create
    - PRODUCT_CREATED
    - PRODUCT_UPDATED
    - PRODUCT_VARIANT_CREATED
    - PRODUCT_VARIANT_UPDATED

2. Upon product update
    - PRODUCT_UPDATED
    - PRODUCT_VARIANT_UPDATED
    - PRODUCT_UPDATED
    - PRODUCT_VARIANT_UPDATED
    
Saleor core version - 3.4.1

**Comment by timuric** [comment]
@satvikt is there a chance you are doing a mutation inside your webhook handlers? Could you please provide saleor version?

**Reply by satvikt** [reply]
@timuric I am **not** doing any mutation inside the webhook handlers.
Saleor version - 3.4.1

**Reply by satvikt** [reply]
@timuric I updated the question above to include all the webhooks fired upon product created/updated.

**Reply by timuric** [reply]
@satvikt this is caused by the dashboard sending a few mutations on save. There are no issues with the core, I will transfer this to dashboard

**Comment by timuric** [comment, marked as answer]
Opened an issue in the dashboard repo https://github.com/saleor/saleor-dashboard/issues/2096

**Reply by satvikt** [reply]
@timuric, thanks for the prompt clarification.

Could you please also take a look at this discussion - https://github.com/saleor/saleor/discussions/9964?
